datafusion/core/tests/user_defined/user_defined_scalar_functions.rs

```diff
@@ -619,6 +624,117 @@ async fn verify_udf_return_type() -> Result<()> {
     Ok(())
 }
 
+#[derive(Debug, Default)]
+struct MockFunctionFactory {
+    pub captured_expr: Mutex<Option<Expr>>,
+}
+
+#[async_trait::async_trait]
+impl FunctionFactory for MockFunctionFactory {
+    #[doc = r" Crates and registers a function from [CreateFunction] statement"]
+    #[must_use]
+    #[allow(clippy::type_complexity, clippy::type_repetition_in_bounds)]
+    async fn create(
+        &self,
+        _config: &SessionConfig,
+        statement: CreateFunction,
+    ) -> datafusion::error::Result<RegisterFunction> {
+        // this function is a mock for testing
+        // `CreateFunction` should be used to derive this function
+
+        let mock_add = Arc::new(|args: &[datafusion_expr::ColumnarValue]| {
+            let args = datafusion_expr::ColumnarValue::values_to_arrays(args)?;
+            let base =
+                datafusion_common::cast::as_float64_array(&args[0]).expect("cast failed");
+            let exponent =
+                datafusion_common::cast::as_float64_array(&args[1]).expect("cast failed");
+
+            let array = base
+                .iter()
+                .zip(exponent.iter())
+                .map(|(base, exponent)| match (base, exponent) {
+                    (Some(base), Some(exponent)) => Some(base.add_wrapping(exponent)),
+                    _ => None,
+                })
+                .collect::<arrow_array::Float64Array>();
+            Ok(datafusion_expr::ColumnarValue::from(
+                Arc::new(array) as arrow_array::ArrayRef
+            ))
+        });
+
+        let args = statement.args.unwrap();
+        let mock_udf = create_udf(
+            &statement.name,
+            vec![args[0].data_type.clone(), args[1].data_type.clone()],
+            Arc::new(statement.return_type.unwrap()),
+            datafusion_expr::Volatility::Immutable,
+            mock_add,
+        );
+
+        // capture expression so we can verify
+        // it has been parsed
+        *self.captured_expr.lock() = statement.params.return_;
+
+        Ok(RegisterFunction::Scalar(Arc::new(mock_udf)))
+    }
+
+    async fn remove(
```

Comment:
Thinking about it more, maybe we don't even need a `remove` in the `FunctionFactory` at all  -- dropping the function is mostly handled by `deregister_udf`?

Or maybe there is some sort of deiniitialization of the function that is needed that would be `FunctionFactory` specific🤔. However, perhaps UDFs that needed that could implement some sort of deregistering / deallocation on `drop`.

Handling deregistering on `drop` might be better anyways as would prevents a function from being deregistered out from under a running execution...

Reply:
lets say we have a SQL UDFs which are expressed by expr, we would need to have some kind of registry (which is not udf_registry) which can be consulted by plan analyser.

so in this case we would have some kind of mock function stored in function registry (which would do nothing, just for planning) and have an expression registry, which would be used to "replace" udf's in logical plan ... 

btw, i believe adding sql udf would be trivial once we get this change and fix few issues with expression namings

Reply:
> lets say we have a SQL UDFs which are expressed by expr, we would need to have some kind of registry (which is not udf_registry) which can be consulted by plan analyser.

Another potential approach would be to encapsulate the code directly into the ScalarUDF and then use an API like `simplify()` that @jayzhan211  and I are working on in https://github.com/apache/arrow-datafusion/pull/9304 to replace the `ScalarUDF` with its `Expr` equivelnt...

Reply:
will have to follow up on #9304 ...  `CREATE FUNCTION` could create UDF or UDaF ... do you maybe have idea how could we decide which unregister_ud*f to call when we get `DROP FUNCTION`?

Reply:
> do you maybe have idea how could we decide which unregister_ud*f to call when we get DROP FUNCTION?

That is a (very) good question. I don't have a good answer at the moment. I'll try and think of one

Reply:
I thought more about this last night and one potential solution is to simply try and deregister all functions with the specified name (udf, udaf, udwf, etc)

It seems non ideal however as it would allow dropping any udf function, not just the ones created by `CREATE FUNCTION` 🤔

Reply:
I think that's good solution. 

On that note there can be UDF and udaf with same name, if used in aggregation query, analyser will freak out, as it will find UDF function. 

Will remove `remove` function from trait for now, we can always put it back

Reply:
`remove` has been gone :), `DROP FUNCTION` will not discriminate against udf type and drop them all.

`remove` is dropped with condition that `simplify` from #9304 find its way to other function types, otherwise function factory would have to implement complicated registry with weak ref or something like that